### PR TITLE
Track wd1

### PR DESCRIPTION
**Pull request by bobhy**
Fires chdir event when OS level change in working directory is detected (based on discussion in #1666).
Also fixes indexerror in cd(['.'] and AUTO_PUSHD on Windows (extension of fix from #1667).

### CODE REVIEW

**Comment by astronouth7303** [comment]
Woo! Thanks for doing this for me.

**Comment by bobhy** [comment]
Could some py.test wizard take a look at e.g, the appveyor failure and help me understand why my test is failing?  It also fails on my machine when I run all the tests together (```py.test```), but does not fail when I run the single module (```py.test tests\test_dirstack_unc.py``). Nor "should" it fail, based on my understanding of the code (famous last words).  

Is there some multitasking hazard in multiple tests accessing the singleton ```dirstack.DIRSTACK``` when run in the test harness?

File: tests/test_dirstack.py [hunk 1]
```diff
@@ -89,3 +89,40 @@ def handler(old, new):
...
+        assert old_dir == os.getcwd()
+        assert old_ds_size + 1 == len(dirstack.DIRSTACK)
+
+
+
+
+
+
+
+
```
Comment (scopatz) [inline tests/test_dirstack.py]: PEP8 excess whitespace.

**Comment by scopatz** [comment]
Hi @bobhy - I ma not a pytest wizard.  But maybe @laerus or @melund can take a look?

File: tests/test_dirstack_unc.py [hunk 2]
```diff
@@ -112,70 +118,70 @@ def test_uncpushd_push_to_same_share(xonsh_builtins):
     assert owd.casefold() == xonsh_builtins.__xonsh_env__['PWD'].casefold()
     dirstack.pushd([r'\\localhost\uncpushd_test_HERE'])
     wd = os.getcwd()
-    assert os.path.splitdrive(wd)[0].casefold() == 'z:'
+    assert os.path.splitdrive(wd)[0].casefold() == TEMP_DRIVE[0]
     assert os.path.splitdrive(wd)[1].casefold() == '\\'
     assert len(_unc_tempDrives) == 1
     assert len(DIRSTACK) == 1
 
     dirstack.pushd([r'\\localhost\uncpushd_test_HERE'])
```
Comment (scopatz) [inline tests/test_dirstack_unc.py]: Is it a problem that you are pushing the same dir name twice?

Comment (scopatz) [inline tests/test_dirstack_unc.py]: That seems like it would cause the len test error below

**Comment by melund** [comment]
>  It also fails on my machine when I run all the tests together (py.test), but does not fail when I run the single module (`py.test tests\test_dirstack_unc.py`). Nor "should" it fail, based on my understanding of the code (famous last words

I think we have a more general issue with our test setup. I had a similar problem when testing `test_*.xsh` files. The always failed when running after certain of the other test files. But they work by them self, or if they are run first. (The workaround fix I used was to always run them first) 

There is a list of files which cause problems here:
https://github.com/xonsh/xonsh/pull/1386#issuecomment-243121021

Are they the same files that cause problems for you @bobhy? 

For examples, does this work? 
`py.test test\test_aliases.py tests\test_dirstack_unc.py`

`test_aliases.py` was one of the files that caused problems for me.

**Comment by laerus** [comment]
any chance that the `__del__` method on `execer.Execer` is the root of all evil?

**Comment by bobhy** [comment]
Thanks for the valuable hint about order of tests, @melund!  I found that I could make my error go away by running test_dirstack_unc.py *before* test_dirstack.py (just those 2 modules).  Closer examination of my changes to `test_dirstack.py` showed I was saving and restoring ```dirstack.DIRSTACK``` around the `yield` in a context manager local to that file.  I don't see how that could possibly have affected tests in a different module , but the failure had to do with ```DIRSTACK``` not being what it "should" have been, and eliminating the context manager from the test got rid of the error. *post hoc, ergo propter hoc.*

Here's the seemingly offending code, if wiser heads want to ponder it.  Maybe there's a similar explanation for the other test failures? 

```python
@contextmanager
def cd_dir(cd_arg, builtins):
    old_dir = os.getcwd()
    old_ds = dirstack.DIRSTACK.copy()
    old_pwd = builtins.__xonsh_env__['PWD']
    dirstack.cd(cd_arg)
    yield
# does python runtime absolutely guarantee that body of `with` block will run immediately
# after this yield statement?  And that the following code will run before any other python code
# on this thread?  
    dirstack.DIRSTACK = old_ds
    os.chdir(old_dir)
    builtins.__xonsh_env__['PWD'] = old_pwd
```

**Comment by scopatz** [comment]
> `# does python runtime absolutely guarantee that body of `with` block will run immediately
> `# after this yield statement?  And that the following code will run after that `with` block
> `# but before any other python code on this thread? 

It does guarantee that tit will be run after the yield on the same thread, with some caveats,

1. Not necessarily immediately, because there may be other context managers in the same with-statement.
2. I think certain exceptions, like SystemExit and KeyboardInterrupt  can break out of this.
3. If an exception happens in another context manager's exit then all bets are off.

That said, py.test does some highly magical things, so I am not sure we can blame this on normal Python.

Also, this passes now, so do we think that it is ready to be merged?

**Comment by bobhy** [comment]
Yes, I'm ready to have it merged.

Thanks,
Bob


On Sep 8, 2016, at 20:07, Anthony Scopatz <notifications@github.com<mailto:notifications@github.com>> wrote:

# does python runtime absolutely guarantee that body ofwithblock will run immediately
# after this yield statement? And that the following code will run after that with block
`# but before any other python code on this thread?

It does guarantee that tit will be run after the yield on the same thread, with some caveats,

  1.  Not necessarily immediately, because there may be other context managers in the same with-statement.
  2.  I think certain exceptions, like SystemExit and KeyboardInterrupt can break out of this.
  3.  If an exception happens in another context manager's exit then all bets are off.

That said, py.test does some highly magical things, so I am not sure we can blame this on normal Python.

Also, this passes now, so do we think that it is ready to be merged?

**Comment by melund** [comment]
I get the following errors locally:

```
================================== FAILURES ===================================
________________________ test_uncpushd_simple_push_pop ________________________

xonsh_builtins = <module 'builtins' (built-in)>
shares_setup = [['\\\\localhost\\uncpushd_test_HERE', 'y:', 'C:\\Users\\mel\\Documents\\xonsh\\tests'], ['\\\\localhost\\uncpushd_test_PARENT', 'w:', 'C:\\Users\\mel\\Documents\\xonsh']]

    @pytest.mark.skipif( not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_simple_push_pop(xonsh_builtins, shares_setup):
        xonsh_builtins.__xonsh_env__ = Env(CDPATH=PARENT, PWD=HERE)

        dirstack.cd([PARENT])
        owd = os.getcwd()
        assert owd.casefold() == xonsh_builtins.__xonsh_env__['PWD'].casefold()
        dirstack.pushd([r'\\localhost\uncpushd_test_HERE'])
        wd = os.getcwd()
>       assert os.path.splitdrive(wd)[0].casefold() == TEMP_DRIVE[0]
E       assert 'c:' == 'z:'
E         - c:
E         + z:

tests\test_dirstack_unc.py:102: AssertionError
---------------------------- Captured stderr setup ----------------------------
This shared resource does not exist.

More help is available by typing NET HELPMSG 2310.

System error 5 has occurred.

Access is denied.

System error 53 has occurred.

The network path was not found.

This shared resource does not exist.

More help is available by typing NET HELPMSG 2310.

System error 5 has occurred.

Access is denied.

System error 53 has occurred.

The network path was not found.

______________________ test_uncpushd_push_to_same_share _______________________

xonsh_builtins = <module 'builtins' (built-in)>

    @pytest.mark.skipif( not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_push_to_same_share(xonsh_builtins):
        xonsh_builtins.__xonsh_env__ = Env(CDPATH=PARENT, PWD=HERE)

        dirstack.cd([PARENT])
        owd = os.getcwd()
        assert owd.casefold() == xonsh_builtins.__xonsh_env__['PWD'].casefold()
        dirstack.pushd([r'\\localhost\uncpushd_test_HERE'])
        wd = os.getcwd()
>       assert os.path.splitdrive(wd)[0].casefold() == TEMP_DRIVE[0]
E       assert 'c:' == 'z:'
E         - c:
E         + z:

tests\test_dirstack_unc.py:118: AssertionError
_____________________ test_uncpushd_push_other_push_same ______________________

xonsh_builtins = <module 'builtins' (built-in)>

    @pytest.mark.skipif( not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_push_other_push_same(xonsh_builtins):
        """push to a, then to b. verify drive letter is TEMP_DRIVE[2], skipping already used TEMP_DRIVE[1]
           Then push to a again. Pop (check b unmapped and a still mapped), pop, pop (check a is unmapped)"""
        xonsh_builtins.__xonsh_env__ = Env(CDPATH=PARENT, PWD=HERE)

        dirstack.cd([PARENT])
        owd = os.getcwd()
        assert owd.casefold() == xonsh_builtins.__xonsh_env__['PWD'].casefold()
        dirstack.pushd([r'\\localhost\uncpushd_test_HERE'])
>       assert os.getcwd().casefold() == TEMP_DRIVE[0] + '\\'
E       assert 'c:\\users\\m...uments\\xonsh' == 'z:\\'
E         - c:\users\mel\documents\xonsh
E         + z:\

tests\test_dirstack_unc.py:147: AssertionError
_______________________ test_uncpushd_cd_unc_auto_pushd _______________________

xonsh_builtins_cd = <module 'builtins' (built-in)>, with_unc_check_enabled = 0

    @pytest.mark.skipif(not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_cd_unc_auto_pushd(xonsh_builtins_cd, with_unc_check_enabled):
        xonsh_builtins_cd.__xonsh_env__['AUTO_PUSHD'] = True
>       so, se, rc = dirstack.cd([r'\\localhost\uncpushd_test_PARENT'])

tests\test_dirstack_unc.py:249:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
xonsh\dirstack.py:208: in cd
    d = _try_cdpath(d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

apath = '\\\\localhost\\uncpushd_test_PARENT'

    def _try_cdpath(apath):
        # NOTE: this CDPATH implementation differs from the bash one.
        # In bash if a CDPATH is set, an unqualified local folder
        # is considered after all CDPATHs, example:
        # CDPATH=$HOME/src (with src/xonsh/ inside)
        # $ cd xonsh -> src/xonsh (whith xonsh/xonsh)
        # a second $ cd xonsh has no effects, to move in the nested xonsh
        # in bash a full $ cd ./xonsh is needed.
        # In xonsh a relative folder is allways preferred.
        env = builtins.__xonsh_env__
        cdpaths = env.get('CDPATH')
>       for cdp in cdpaths:
E       TypeError: 'NoneType' object is not iterable

xonsh\dirstack.py:166: TypeError
________________________ test_uncpushd_cd_unc_nocheck _________________________

xonsh_builtins_cd = <module 'builtins' (built-in)>, with_unc_check_disabled = 0

    @pytest.mark.skipif(not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_cd_unc_nocheck(xonsh_builtins_cd, with_unc_check_disabled):
>       dirstack.cd([r'\\localhost\uncpushd_test_HERE'])

tests\test_dirstack_unc.py:258:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
xonsh\dirstack.py:208: in cd
    d = _try_cdpath(d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

apath = '\\\\localhost\\uncpushd_test_HERE'

    def _try_cdpath(apath):
        # NOTE: this CDPATH implementation differs from the bash one.
        # In bash if a CDPATH is set, an unqualified local folder
        # is considered after all CDPATHs, example:
        # CDPATH=$HOME/src (with src/xonsh/ inside)
        # $ cd xonsh -> src/xonsh (whith xonsh/xonsh)
        # a second $ cd xonsh has no effects, to move in the nested xonsh
        # in bash a full $ cd ./xonsh is needed.
        # In xonsh a relative folder is allways preferred.
        env = builtins.__xonsh_env__
        cdpaths = env.get('CDPATH')
>       for cdp in cdpaths:
E       TypeError: 'NoneType' object is not iterable

xonsh\dirstack.py:166: TypeError
_____________________ test_uncpushd_cd_unc_no_auto_pushd ______________________

xonsh_builtins_cd = <module 'builtins' (built-in)>, with_unc_check_enabled = 0

    @pytest.mark.skipif(not ON_WINDOWS, reason="Windows-only UNC functionality")
    def test_uncpushd_cd_unc_no_auto_pushd(xonsh_builtins_cd, with_unc_check_enabled):
>       so, se, rc = dirstack.cd([r'\\localhost\uncpushd_test_PARENT'])

tests\test_dirstack_unc.py:264:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
xonsh\dirstack.py:208: in cd
    d = _try_cdpath(d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

apath = '\\\\localhost\\uncpushd_test_PARENT'

    def _try_cdpath(apath):
        # NOTE: this CDPATH implementation differs from the bash one.
        # In bash if a CDPATH is set, an unqualified local folder
        # is considered after all CDPATHs, example:
        # CDPATH=$HOME/src (with src/xonsh/ inside)
        # $ cd xonsh -> src/xonsh (whith xonsh/xonsh)
        # a second $ cd xonsh has no effects, to move in the nested xonsh
        # in bash a full $ cd ./xonsh is needed.
        # In xonsh a relative folder is allways preferred.
        env = builtins.__xonsh_env__
        cdpaths = env.get('CDPATH')
>       for cdp in cdpaths:
E       TypeError: 'NoneType' object is not iterable

xonsh\dirstack.py:166: TypeError
=========================== pytest-warning summary ============================
WC1 None [pytest] section in setup.cfg files is deprecated, use [tool:pytest] instead.
==== 6 failed, 2694 passed, 13 skipped, 1 pytest-warnings in 90.68 seconds ====
Exception ignored in: <bound method Execer.__del__ of <xonsh.execer.Execer object at 0x0000018509A354A8>>
Traceback (most recent call last):
  File "c:\users\mel\documents\xonsh\xonsh\execer.py", line 46, in __del__
  File "c:\users\mel\documents\xonsh\xonsh\built_ins.py", line 1000, in unload_builtins
  File "C:\Users\mel\Anaconda3\lib\abc.py", line 187, in __instancecheck__
AttributeError: 'NoneType' object has no attribute '_abc_invalidation_counter'
```

**Comment by melund** [comment]
Oh. Sorry. The error were also there on master.

**Pull request merged**